plugin/pki/path_venafi_cert_enroll.go

```diff
@@ -157,76 +166,161 @@ func (b *backend) pathVenafiSign(ctx context.Context, req *logical.Request, data
+		b.Logger().Info(fmt.Sprintf("Calling prevent local for hash %v", certId))
 		logicalResp := preventReissueLocal(b, ctx, logicalRequest, reqData, role, certId)
 		if logicalResp != nil {
 			return logicalResp, nil
 		}
 	}
 
+	// When utilizing performance standbys in Vault Enterprise, this forces the call to be redirected to the primary since
+	// a storage call is made after the API calls to issue the certificate.  This prevents the certificate from being
+	// issued twice in this scenario.
+	if !role.NoStore && b.System().ReplicationState().
+		HasState(consts.ReplicationPerformanceStandby|consts.ReplicationPerformanceSecondary) {
+		return nil, logical.ErrReadOnly
+	}
+
+	// if user is using store by hash
+	var cert *SyncedResponse
+	if !signCSR && role.StoreBy == storeByHASHstring {
+		found := false
+		b.Logger().Info("locking process to update cache")
+		b.mux.Lock()
+		cert, found = cache[certId]
+		if found {
+			b.Logger().Info(fmt.Sprintf("Request is waiting on previous request on certificate to be issued for hash %v", certId))
+			cert.condition.Wait()
+			b.mux.Unlock()
+			b.Logger().Info(fmt.Sprintf("Returning the certificate for hash %v retrieved from previous request.", certId))
+			if cert.error != nil {
+				b.Logger().Error("waiting state error: %s", cert.error.Error())
+				return nil, cert.error
+			}
+			return cert.logResponse, nil
+		}
+		newCond := sync.NewCond(&b.mux)
+		cert = &SyncedResponse{
+			logResponse: nil,
+			condition:   newCond,
+			error:       nil,
+		}
+		cache[certId] = cert
+		b.mux.Unlock()
+	}
+
 	var certReq *certificate.Request
 	certReq, err = formRequest(*reqData, role, &connector, signCSR, b.Logger())
 	if err != nil {
+		if !signCSR && role.StoreBy == storeByHASHstring {
+			b.recoverBroadcast(cert, logResp, certId, err)
+		}
+		b.Logger().Error("error error forming request: %s", err.Error())
 		return logical.ErrorResponse(err.Error()), nil
 	}
 
 	err = createCertificateRequest(b, &connector, ctx, logicalRequest, role, certReq)
 	if err != nil {
+		if !signCSR && role.StoreBy == storeByHASHstring {
+			b.recoverBroadcast(cert, logResp, certId, err)
+		}
+		b.Logger().Error("error creating certificate request: %s", err.Error())
 		return nil, err
 	}
 	var pcc *certificate.PEMCollection
 	pcc, err = runningEnrollRequest(b, data, certReq, connector, role, signCSR)
 	if err != nil {
+		if !signCSR && role.StoreBy == storeByHASHstring {
+			b.recoverBroadcast(cert, logResp, certId, err)
+		}
+		b.Logger().Error("error running enroll request: %s", err.Error())
 		return nil, err
 	}
 
-	var logResp *logical.Response
 	logResp, err = b.storingCertificate(ctx, logicalRequest, pcc, role, signCSR, certId, (*reqData).commonName, (*reqData).keyPassword)
+	if err != nil {
+		if !signCSR && role.StoreBy == storeByHASHstring {
+			b.recoverBroadcast(cert, logResp, certId, err)
+		}
+		b.Logger().Error("error storing certificate: %s", err.Error())
```

Comment:
If cert is not stored properly, but caller is not using store by at all, we can ignore and not send error right ?

Reply:
If the user is not using any store_by, e.g. setting `role.NoStore == true`, then yes, we will not have error at all so if would not go into that statement